### PR TITLE
[SGD] v2 alpha: Tensorflow Backend

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

Implement `Tensorflow` as a backend for SGD `Trainer`.

Implementation notes:
1. Currently there are no user defined properties in `TensorflowConfig`, but can be added in the future.
2. This works with `MultiWorkerMirroredStrategy`. Further adjustments may be needed to support additional strategies.
3. The behavior when `num_workers=1` is not well defined. If `training_func` tries to read from `os.environ["TF_CONFIG"]` they will run into an error.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by amogkam — APPROVED** [review]
LGTM, just left some minor comments.

File: python/ray/util/sgd/v2/backends/tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+@dataclass
+class TensorflowConfig(BackendConfig):
+    @property
+    def backend_cls(self):
+        return TensorflowBackend
+
+
+def setup_tensorflow_process_group(worker_addresses, index):
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/util/sgd/v2/backends/tensorflow.py:21]: Also per our discussion rename `setup_tensorflow_process_group`.

[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/util/sgd/v2/backends/tensorflow.py]: - added type hints
- updated `index` doc to mention `world_rank`
- renamed to `setup_tensorflow_environment`

File: python/ray/util/sgd/v2/backends/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                        worker_addresses=urls,
+                        index=i))
+            ray.get(setup_futures)
+
+        else:
+            logger.info("Distributed Tensorflow is not being used.")
+
+    def on_shutdown(self, worker_group: WorkerGroup,
+                    backend_config: BackendConfig):
+        pass
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/util/sgd/v2/backends/tensorflow.py:71]: Added a comment mentioning there's currently nothing to shutdown.

File: python/ray/util/sgd/v2/tests/test_trainer.py [hunk 3]
```diff
@@ -64,6 +66,28 @@ def train_func(config):
     assert all(result == "banana" for result in results)
 
 
+def test_tensorflow_mnist(ray_start_2_cpus):
```
Comment (richardliaw) [inline python/ray/util/sgd/v2/tests/test_trainer.py:69]: add a gpu test (skip if no gpus available)?

File: python/ray/util/sgd/v2/examples/tensorflow_mnist_example.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    args, _ = parser.parse_known_args()
+
+    import ray
+
+    if args.smoke_test:
+        ray.init(num_cpus=2)
+    elif args.server_address:
+        ray.util.connect(args.server_address)
+    else:
+        ray.init(address=args.address)
```
Comment (richardliaw) [inline python/ray/util/sgd/v2/examples/tensorflow_mnist_example.py:118]: we just got rid of util.connect btw

**Pull request merged**